feat(menu): Submenu

jurokapsiar opened the pull request:
Based on two previous menu/submenu PRs. Fixes https://github.com/stardust-ui/react/issues/542

Missing things:
- [x] document click / focus out should close submenu
- [x] action should close submenu
- [x] left/right key should move to the next/previous menu item according to ARIA practices
- [x] submenu indicator
- [x] commented out overflow, not sure why it was there, need to verify
- [x] verify wrapper render

Update by @mnajdova 
Some questions:
- I needed to add key to the wrapper element created, so in the defaultProps I added key that is generated by a new added dependency `uniqid`. Are we ok with this?
- The tests are failing, I guess because I added `Ref` wrapper around the `ElementType`. How can be tackle this? Is the wrapper maybe causing this?
- How are we handling event handlers when we have wrapper? There is issue with the onClick event, which is added on the wrapper, but on the anchor is added only if the wrapper is not added. Am I doing something wrong here?

The questions were resolved thanks to @layershifter

All accessibility requirements are added, let me know if you find some issue with it.

Commit:
feat(menuItem: add menu prop)

Commit:
Only one submenu open at a time

Commit:
Fix bug: submenu noe closes on clicking a menuItem with no submenu

Commit:
Code cleanup

Commit:
Remove the activeIndex prop passed to the menuItem

Commit:
Remove Popup and implement submenu without it

jurokapsiar committed:
Merge branch 'master' into feat/menuitem-menu-prop

jurokapsiar committed:
initial keyboard support

jurokapsiar committed:
left/right arrow handling

jurokapsiar committed:
simplify keyboard handlers

jurokapsiar commented on src/themes/teams/components/Menu/menuStyles.ts:
```diff
@@ -28,7 +28,7 @@ export default {
           }),
           borderRadius: pxToRem(4),
-          overflow: 'hidden',
+          // overflow: 'hidden',
```
this needs to be verified - it breaks submenu when oveflow is hidden

mnajdova replied:
This is the PR that added it: https://github.com/stardust-ui/react/pull/360 Will sync with @Bugaa92 for finding alternative solution for the cropped borders, I have new ideas.

jurokapsiar commented on src/themes/teams/components/Menu/menuVariables.ts:
```diff
@@ -24,7 +24,7 @@ export interface MenuVariables {
   return {
     defaultColor: siteVars.gray02,
-    defaultBackgroundColor: 'transparent',
+    defaultBackgroundColor: '#FFF',
```
this should come from theme/site

Commit:
-fixed import in the MenuItem

Commit:
-added submenuIndicator
-small fixes in the styles and the way the submenu is generated

Commit:
-clicking on leaf element should close the submenus (the same should be done for enter/space)
-applied consistent (left-right) navigation for horizontal menu and (up-down) navigation for vertical menu

Commit:
Merge branch 'master' into feat/menuitem-menu-prop

# Conflicts:
#	src/components/Menu/Menu.tsx
#	src/components/Menu/MenuItem.tsx

Commit:
-implemented outside click to close all menus
-implemented enter key on leaf menu item to close the menu

Commit:
-fixed import

sophieH29 commented on src/components/Menu/MenuItem.tsx:
```diff
@@ -176,6 +272,48 @@ class MenuItem extends UIComponent<Extendable<MenuItemProps>, MenuItemState> {
+      this.trySetState({ submenuOpen: false })
+    }
+    _.invoke(props, 'onKeyDown', props)
+  }
```
Keys should be assigned to actions in behaviors. I think we can have one action called ```performSubmenuClick``` which will call ```handleSubmenuClicked``` on Enter or Space. Same way as we have on MenuItem ```performClick``` line 231.

So basically in ```menuItemBehavior``` you add to key action

```js
keyActions: {
   ...
   
   submenu: {
      performSubmenuClick: {
        keyCombinations: [{ keyCode: keyboardKey.Enter }, { keyCode: keyboardKey.Spacebar }],
      },
    },
   ...
}
```

In ```MenuItem``` comp, add action handler (line 231)

```js
actionHandlers = {
    ...
    performSubmenuClick: (event) => this.handleSubmenuClicked(event, null)

    ...
}
```

NOTE: I see that in your handlers you receive props as second param. In case of ```keydown``` event, client code is handled in```getKeyDownHandlers.ts```, by calling ```_.invoke(props, 'onKeyDown', event)```, so we can just pass NULL as 2nd param in action handler.

and then line 188 where you create submenu

```js
const maybeSubmenu =
       ...
        Menu.create(menu, {
            defaultProps: {
              accessibility: submenuBehavior, 
              
            },
            overrideProps: {
              onClick: this.handleSubmenuClicked,
             ...accessibility.keyHandlers.submenu   // adds onkeyDown handler which will call 'performSubmenuClick'
            },
          })
        : null
```

Please let me know what you think.

Commit:
-refactored MenuItem handlers - fixed issues
-removed onClick handler for the Menu (not necessary for now)
-added onKeyDown in the creation of the MenuItem in the Menu component for handling the action prop

Commit:
-added setActiveIndex callback and removed onKeyDown in the creation of MenuItems in the Menu component

Commit:
-right arrow key is closing the submenus and goes to the next element if the menu is horizontal, or is focusing the first MenuItem if it is vertical

Commit:
-handled left arrow key

Commit:
-changed ref
-focus trap wip

Commit:
-added Ref component instead of using the itemRef on the ElementType
-removed subscription for focus

Commit:
Merge branch 'master' into feat/menuitem-menu-prop

# Conflicts:
#	src/components/Menu/MenuItem.tsx
#	src/lib/getKeyDownHandlers.ts

Commit:
-fixes

Commit:
-moved ref //TODO: figure out tests failing

Commit:
-close menu on outside focus

Commit:
-improved comments

Commit:
-fixed escape key not focusing the active element
-changed parentRef to inSubmenu boolean

Commit:
-fixing key problems
-added dependency for generating id

Commit:
-refactored submenuRef element

layershifter committed:
fix broken tests

layershifter commented on src/components/Menu/MenuItem.tsx:
```diff
@@ -118,32 +152,74 @@ class MenuItem extends UIComponent<Extendable<MenuItemProps>, MenuItemState> {
+        <Ref
+          innerRef={domElement => {
+            this.submenuDomElement = domElement
+          }}
```
Please use a `createRef()` API there: https://github.com/stardust-ui/react/pull/584#discussion_r241463373

mnajdova replied:
Done, thanks!

Commit:
-added comments in the tests
-changed the submenuDomElement so submenuRef

Commit:
Merge branch 'master' into feat/menuitem-menu-prop

Commit:
-renamed inSubmenu to submenu prop in the Menu

Commit:
-fixed with the auto-controlled prop in the Menu

Commit:
-added state interface in the Menu
-improved menu variables' names

Commit:
-fixed variables in examples

miroslavstastny committed:
chore: prepare release 0.15.0 [ci skip]

kuzhelov commented on src/components/Menu/MenuItem.tsx:
```diff
@@ -153,18 +225,69 @@ class MenuItem extends UIComponent<Extendable<MenuItemProps>, MenuItemState> {
+    this.outsideClickSubscription.unsubscribe()
+
+    if (this.props.menu && this.state.submenuOpen) {
+      setTimeout(() => {
```
note, this approach (presumably, taken from the Popup) has one subtle problem: although it solves the issue of item being opened and closed within the same frame, it also opens possibility for the following behavior:

- MenuItem has opened submenu, this opened prop value doesn't change anyhow for the next JS frame.

///------ JS FRAME --------////
- render method was called (but `submenuOpen ` wasn't changed changed anyhow, it is still `true`)
- updateOutsideClickSubscription
  -  this.outsideClickSubscription.unsubscribe(), no subscription on outside click (!!)
  - setTimeout(() => subscription will be made on the next frame)

- ... some other logic ..
   - note, there is no click subscription at this moment, but it should be (!!)

////---------------------------////

Would suggest to leave a fix for a separate PR, for both components.

mnajdova replied:
Will create issue for this, after the PR will be merged

miroslavstastny committed:
0.15.0

kuzhelov commented on src/components/Menu/MenuItem.tsx:
```diff
@@ -110,6 +139,11 @@ class MenuItem extends UIComponent<Extendable<MenuItemProps>, MenuItemState> {
     vertical: PropTypes.bool,
     wrapper: PropTypes.oneOfType([PropTypes.node, PropTypes.object]),
+    menu: customPropTypes.itemShorthand,
+    submenuOpen: PropTypes.bool,
```
this one is easy to be confused with `openSubmenu` which we have in the same file. Verb, probably, would help to better differentiate - or past tense could be used. Lets see whether this is consistent with the rest of the code (and naming used), but I would rather like to eliminate this confusion.

Potential suggestions:
- `isSubmenuOpen`
- `submenuOpened`

mnajdova replied:
We never use `is` as a prefix for the boolean props, and the similar auto-controlled prop in the Popup, has the name `open`. But, I agree with your concerns, would rather rename it to `submenuOpened`, but not sure as it won't be consistent. What do you think?

kuzhelov replied:
lets rely on consistency first - by that we will ensure that problem will be fixed for all components and in the same way. If this name is consistent with other components' prop names, lets leave it 👍

kuzhelov commented on src/components/Menu/MenuItem.tsx:
```diff
@@ -118,32 +152,70 @@ class MenuItem extends UIComponent<Extendable<MenuItemProps>, MenuItemState> {
+
   state = {
     isFromKeyboard: false,
+    submenuOpen: false,
```
won't this override the effect of `submenuOpen={true}` prop being provided by the client? Have taken a look on the logic of `AutoControlledComponent` (`getAutoControlledStateValue`) and it seems to be pretty much the case

mnajdova replied:
Yep, we should never have initial value for e state that is part of the Autocontrolled prop. This was leftover from the prev PRs merged here. Will remove it. Thanks for the catch!

kuzhelov commented on docs/src/examples/components/Menu/Types/MenuExampleWithSubmenu.shorthand.tsx:
```diff
@@ -0,0 +1,57 @@
+        {
+          key: '2',
+          content: 'item2',
+          menu: { items: [{ key: '1', content: 'item2.1' }, { key: '2', content: 'item2.2' }] },
```
👍 for naming consistency from level to level in the nested tree (no props like `submenu`, etc)

Commit:
-remove state initialization in the MenuItem component

kuzhelov commented on src/components/Menu/MenuItem.tsx:
```diff
@@ -153,18 +225,69 @@ class MenuItem extends UIComponent<Extendable<MenuItemProps>, MenuItemState> {
         }),
       })
     }
     return menuItemInner
```
sorry, just to understand what current limitations are. What will happen if `wrapper` will be evaluated to falsy value?
There won't be any submenu functionality supported in this case, right?

mnajdova replied:
Yes, because the submenu as an element is part of the wrapper (it cannot be part of the anchor tag, which is the element created if the wrapper is not provided). I agree that this is not obvious, maybe we should add that in the doc somewhere.

kuzhelov commented on src/components/Menu/MenuItem.tsx:
```diff
@@ -118,32 +152,65 @@ class MenuItem extends UIComponent<Extendable<MenuItemProps>, MenuItemState> {
+          onFocus={this.handleFocus}
+          {...accessibility.attributes.anchor}
+          {...rest}
+          {...!wrapper && { onClick: this.handleClick }}
```
are we really sure that we would like to override client's `onClick` logic in that case?

mnajdova replied:
Actually, we are not overriding. The implementation of `handleClick` is the following:
```
  private handleClick = e => {
    this.performClick(e)
    _.invoke(this.props, 'onClick', e, this.props)
  }
```

kuzhelov commented on src/components/Menu/MenuItem.tsx:
```diff
@@ -118,32 +152,65 @@ class MenuItem extends UIComponent<Extendable<MenuItemProps>, MenuItemState> {
+          {Menu.create(menu, {
+            defaultProps: {
+              accessibility: submenuBehavior,
+              vertical: true,
```
is there any possibility that it could be horizontal one? I mean, I might imagine it (horizontal menu and horizontal submenu) - this was the case for Word's toolbar, for example

mnajdova replied:
According to the spec: https://github.com/stardust-ui/react/issues/542 the submenus should always be vertical. @jurokapsiar opinion on this?

kuzhelov replied:
as a proof for my point:

![image](https://user-images.githubusercontent.com/9024564/50227500-27481280-03a6-11e9-8bed-6a37f9a2a812.png)

mnajdova replied:
Will be added in a follow up PR.

sophieH29 commented on docs/src/examples/components/Menu/Types/index.tsx:
```diff
@@ -19,6 +19,16 @@ const Types = () => (
+      examplePath="components/Menu/Types/MenuExampleWithSubmenu"
+    />
+    <ComponentExample
+      title="Vertical submenus"
```
I would name it ```Vertical menu with submenus```. Because both of the examples are vertical submenus :)

sophieH29 commented on src/components/Menu/MenuItem.tsx:
```diff
@@ -179,6 +297,44 @@ class MenuItem extends UIComponent<Extendable<MenuItemProps>, MenuItemState> {
+    const { submenuOpen } = this.state
+    if (menu && submenuOpen) {
+      this.setState({ submenuOpen: false })
+      if (!inSubmenu && (keyboardKey.getCode(e) === keyboardKey.Escape || this.props.vertical)) {
```
what is a use case to have an additional check on ```Esc``` button? We should avoid using keys in components, if you need a specific thing to happen on Esc, better create a separate action handler for that.

mnajdova replied:
Added new handler, thanks for the suggestion!

sophieH29 commented on src/lib/accessibility/Behaviors/Menu/menuBehavior.ts:
```diff
@@ -21,6 +22,8 @@ const menuBehavior: Accessibility = (props: any) => ({
       isCircularNavigation: true,
       preventDefaultWhenHandled: true,
       shouldFocusFirstElementWhenReceivedFocus: true,
+      // TODO: check if this should be applied to other behaviors as well
```
probably, we don't need this comment here :)

mnajdova replied:
I left it intentionally for your guys, as I don't think we had final decision whether we wanted to have this in the other behaviors.

mnajdova replied:
@sophieH29 @jurokapsiar should we add the direction prop to the toolbarBehavior, tabListBehavior as well, cus we have circular navigation there as well?

sophieH29 replied:
yes, we probably will add it in the future when needed. But don't see any reason to have this comment here.

layershifter commented on src/components/Menu/MenuItem.tsx:
```diff
@@ -153,18 +220,69 @@ class MenuItem extends UIComponent<Extendable<MenuItemProps>, MenuItemState> {
+    if (
+      this.itemRef &&
+      (!this.itemRef.current || !this.itemRef.current.contains(e.target)) &&
+      (!this.submenuRef.current || !this.submenuRef.current.contains(e.target))
```
We have a `doesNodeContainClick()` util.

mnajdova replied:
Changed, thanks!

sophieH29 committed:
fix(Prototype): Fix Popover prototype after breaking changes (#623)

Commit:
-added new handler for escape
-changed submenu examples titles

Commit:
-refactored conditions using doesNodeContainClick

kuzhelov commented on docs/src/examples/components/Menu/Types/MenuExampleWithSubmenu.shorthand.tsx:
```diff
@@ -0,0 +1,57 @@
+      componentStyles: {
+        Menu: {
+          root: {
+            zIndex: 1000,
```
shouldn't it be part of the default styles? like

```
// menuItemStyles
'& .ui-menu {
  z-index: ..
}'
```

mnajdova replied:
I thought this was added only because of the docs experience, but it turns out that it is common practice to add z-index on the submenus. Will add this styles in the MenuItem, thanks!

kuzhelov commented on src/components/Menu/Menu.tsx:
```diff
@@ -88,6 +96,7 @@ class Menu extends AutoControlledComponent<Extendable<MenuProps>, any> {
     secondary: customPropTypes.every([customPropTypes.disallow(['primary']), PropTypes.bool]),
     underlined: PropTypes.bool,
     vertical: PropTypes.bool,
+    submenu: PropTypes.bool,
```
here is another clear example - lets just take a look on the following code:

```
<Menu submenu={...} />
```

given this code, what we might expect to be passed as a `submenu`'s value? actually, I would expect items collection. This is what we have for `MenuItem`'s `menu`, but here - this is a boolean flag :)

mnajdova replied:
Totally agree, for consistency I will leave it as this, and will create issue for renaming all boolean props to have the `is` prefix. Thanks for pointing this out.

kuzhelov commented on src/components/Menu/MenuItem.tsx:
```diff
@@ -78,16 +91,32 @@ export interface MenuItemProps
+  menu?: ShorthandValue
+
+  /** Indicates if the submenu is open */
+  submenuOpen?: boolean
```
don't like this names duality - on the one side we have `menu`, on the other one `submenuOpen` - while, in fact, the same exact menu is implied by both. Consider the following expression as an example:

```if (menu && submenuOpen) { ... }```

 Suggest to rename to `menuOpen`

kuzhelov commented on src/components/Menu/MenuItem.tsx:
```diff
@@ -179,6 +297,55 @@ class MenuItem extends UIComponent<Extendable<MenuItemProps>, MenuItemState> {
+    const { menu, inSubmenu } = this.props
+    const { submenuOpen } = this.state
+    if (menu && submenuOpen) {
+      this.setState({ submenuOpen: false })
```
how focus is passed in case if it is not a top-level menu? Are we implicitly relying on event bubbling here?

mnajdova replied:
Fow now yes, let's consider implementing this with callback mechanism in a follow up PR.

kuzhelov commented on src/components/Menu/MenuItem.tsx:
```diff
@@ -110,6 +139,11 @@ class MenuItem extends UIComponent<Extendable<MenuItemProps>, MenuItemState> {
+    menu: customPropTypes.itemShorthand,
+    submenuOpen: PropTypes.bool,
+    defaultSubmenuOpen: PropTypes.bool,
+    setActiveIndex: PropTypes.func,
```
`index` is only used for making this item `active` by this callback. We can remove `index` altogether and rename this prop to something that is more specific (and better reflects intent of this method).

Actually, the thing we have here is event that is fired in controlled mode of `MenuItem` - and, thus, we can name it as `onActiveChanged` and call when `MenuItem` would like to 'activate' itself

kuzhelov reviewed (commented):
lets discuss mentioned points and try to simplify the approach a bit. 

Also, please, note that currently we have the following problem related to visual representation of focused item:

![image](https://user-images.githubusercontent.com/9024564/50117291-da9cf400-024c-11e9-93c3-180b7930101e.png)

Could you say which one os focused now? :)

kuzhelov committed:
chore: cache results of vulnerability scans (#621)

* implement caching strategy

* adjust file name of scan marker

* add yarn lock hash to marker file name

* add change to build config

* fix dir name in build config

* improve caching strategy

* just restore cache

* temporary remove lint and tests

* try

* fix caching strategy

* try

* try

* try

* try epoch

* create file on scan

* return lint and test steps

* introduce comment for the caching approach taken

* remove unnecessary function

* simplify expression for marker file name

Commit:
feat(text): color prop (#597)

* feat(text): color prop

* addressed comments

* changelog

* amended changelog

* made text color override other props that change color

Commit:
-renamed submenu* props to menu in the MenuItem component
-changed setActiveIndex with onActiveChanged
-introduced different styles for the hovering vs active elements

layershifter commented on docs/src/examples/components/Menu/Types/index.tsx:
```diff
@@ -19,6 +19,16 @@ const Types = () => (
+    <ComponentExample
+      title="Vertical menu with submenus"
+      description="A vertical menu can have submenu."
+      examplePath="components/Menu/Types/MenuExampleVerticalWithSubmenu"
```
This example should be present, but we can leverage the attention to it there.

```diff
<ComponentExample
-  title="Vertical menu with submenus"
-  description="A vertical menu can have submenu."
  examplePath="components/Menu/Types/MenuExampleVerticalWithSubmenu"
/>
```

<details>
  <summary>Preview</summary>

![image](https://user-images.githubusercontent.com/14183168/50161705-c99bc380-02dc-11e9-848e-028263e23d70.png)
</details>

layershifter commented on src/components/Menu/Menu.tsx:
```diff
@@ -60,12 +61,19 @@ export interface MenuProps extends UIComponentProps, ChildrenComponentProps {
+  submenu?: boolean
+}
+
+export interface MenuState {
```
Should we export this interface?

kuzhelov replied:
while understand the reason for this sentiment, it turns out that, actually, we should - as in the style functions we need to provide both state and props types, as one of the args of this style function is the result of `props` being merged to `state`:

```js
// this is the code that will be written by client
const buttonStyles = {
  // here props are of type (ButtonState & ButtonProps)
  root: ({ props, ... }) => { ... }
```

mnajdova replied:
Agreed, will export it.

mnajdova replied:
Exported

layershifter commented on src/components/Menu/Menu.tsx:
```diff
@@ -119,11 +137,14 @@ class Menu extends AutoControlledComponent<Extendable<MenuProps>, any> {
-      MenuItem.create(item, {
+    return _.map(items, (item, index) => {
+      const active =
+        typeof activeIndex === 'string' ? parseInt(activeIndex, 10) : activeIndex === index
```
This condition is not valid, if we `activeIndex` is `string` it will be converted to `number`, in other cases it will be boolean:

```js
let activeIndex = 1
let index = 1
let active

active = typeof activeIndex === 'string' ? parseInt(activeIndex, 10) : activeIndex === index
console.log(typeof active) // boolean

activeIndex = '1'
active = typeof activeIndex === 'string' ? parseInt(activeIndex, 10) : activeIndex === index
console.log(typeof active) // number
```

mnajdova replied:
Yep, I am missing brackets
` typeof activeIndex === 'string' ? parseInt(activeIndex, 10) : activeIndex === index` => ` (typeof activeIndex === 'string' ? parseInt(activeIndex, 10) : activeIndex) === index
`

mnajdova replied:
The active prop should always be boolean. Fixed.

Commit:
feat(header): header and header description color prop (#628)

* feat(header): header and header description color prop

* changelog

* fixed examples

* addressed PR comments

Commit:
-improved example
-fixed issue with the condition for the active prop

Commit:
-exported MenuState
-added correct typings to the menuStyles

Commit:
-fixed underlined active + hovered style

kuzhelov committed:
fix(Popup): allow to 'detach' from trigger and RTL adjustments (#612)

* introduce offset prop

* correct description of supported values

* update changelog

* introduce fix

* ensure RTL is properly applied to complex offset expressions

* rename method to make logic more expressive

* add unit tests

* remove unnecessary grid props from offset example

* update changelog

sophieH29 committed:
fix(SelectableList): Items in list should be selectable (#566)

* Reflect which item is selected in list

* Make list derived from autocontrolled component

* small fix

* Update ListExampleSelection.tsx

* Update ListExampleSelection.shorthand.tsx

* Small improvement

* Rename *ItemIndex -> *Index

* Names refactoring

* Minor improvements

* update changelog

* Add onSelectedIndexChange

* Add some tests

* Small improvements afer CR

* Small improvements afer CR

* Small improvements afer CR

* create focus handler when List is constructed

* fix changelog

* changelog

Commit:
-fixed border corner clipped by adding custom styles for the first child menu items and the last child menu items in vertical menu

layershifter committed:
docs(Examples): allow to use TS in examples (#617)

* docs(Examples): allow to use TS in examples

* add jsdoc

* fix typo

* rename file

* add comment

* `createExample` to `createExampleSourceCode`

* rework with `path.relative()`

* remove JSON files on remove tsx

* create getRelativePathToSource function

kuzhelov commented on docs/src/examples/components/Menu/Variations/MenuExampleIconOnlyPrimaryInverted.shorthand.tsx:
```diff
@@ -14,9 +14,9 @@ const MenuExampleIconOnlyPrimaryInverted = () => (
-      defaultColor: siteVars.gray06,
-      defaultBackgroundColor: siteVars.brand,
-      typePrimaryActiveBorderColor: siteVars.white,
+      color: siteVars.gray06,
```
👍

kuzhelov commented on src/components/Menu/MenuItem.tsx:
```diff
@@ -153,18 +220,69 @@ class MenuItem extends UIComponent<Extendable<MenuItemProps>, MenuItemState> {
+      !doesNodeContainClick(this.itemRef.current, e) &&
+      !doesNodeContainClick(this.menuRef.current, e)
+    ) {
+      this.state.menuOpen && this.trySetState({ menuOpen: false })
```
nit: we can make this check (`this.state.menuOpen`) as the first step, because (`doesNodeContainClick`) are less efficient ones:

```js
if (!this.state.menuOpen) {
  return 
}
...
```

kuzhelov commented on src/components/Menu/MenuItem.tsx:
```diff
@@ -153,18 +220,69 @@ class MenuItem extends UIComponent<Extendable<MenuItemProps>, MenuItemState> {
+    const { active, menu } = this.props
+    if (menu) {
+      if (doesNodeContainClick(this.menuRef.current, e)) {
+        // submenu was clicked => close it and propagate
```
thanks for comments, very helpful ones!

kuzhelov commented on src/components/Menu/MenuItem.tsx:
```diff
@@ -179,6 +297,55 @@ class MenuItem extends UIComponent<Extendable<MenuItemProps>, MenuItemState> {
+    const { menuOpen } = this.state
+    if (menu && menuOpen) {
+      this.setState({ menuOpen: false })
+      if (!inSubmenu) {
```
for some cases we have this focus logic being run in the callback of the state change (which is correct), somewhere not. Lets make approach to that being consistent (or, if there is any reason why it is not, please, just let me know)

mnajdova replied:
Initially I had here check for which key was pressed, and I could not use the event in async manner there. That is not the case anymore, as we decided to use different handlers for different keys. We can move this safely in the callback now.

kuzhelov commented on src/components/Menu/MenuItem.tsx:
```diff
@@ -179,6 +297,55 @@ class MenuItem extends UIComponent<Extendable<MenuItemProps>, MenuItemState> {
     _.invoke(this.props, 'onFocus', e, this.props)
   }
+
+  private closeAllMenus = e => {
```
it is hard to infer the difference of logic provided by this method vs subsequent `closeMenuAndFocusNextParentItem` one by using their names only - more than that, quite a lot of implementation code is similar. Maybe we could refactor it to be a single method with some parameter that will drive (and describe) this difference?

Commit:
-addressed comments on PR

Commit:
-updated changelog

Commit:
Merge branch 'master' into feat/menuitem-menu-prop

# Conflicts:
#	CHANGELOG.md

Commit:
-updated changelog

Commit:
-updated changelog with breaking changes
-removed TODO comment

Commit:
Merge branch 'master' into feat/menuitem-menu-prop

# Conflicts:
#	CHANGELOG.md

Commit:
Merge branch 'master' into feat/menuitem-menu-prop

Commit:
-fixed imports in examples
